### PR TITLE
Blog: Dynamic Scaling and Performance

### PR DESCRIPTION
**Pull request by oleschoenburg**
This combines three experiments @deepthidevaki  and I did today, all relating to dynamic scaling and performance.

Closes https://github.com/zeebe-io/zeebe-chaos/issues/456
Closes https://github.com/zeebe-io/zeebe-chaos/issues/457
Closes https://github.com/camunda/zeebe/issues/15517

### CODE REVIEW

**Review by deepthidevaki — APPROVED** [review]
:rocket: Thanks a lot for writing this up.

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+This results in 400 jobs created and completed per second.
+
+This stresses the 3 brokers and we see backpressure on all partitions.
+![](./perf_initial_backpressure.png)
+
+We also see a few jobs timing out, indicating that the cluster is unable to handle this load consistenly:
+![](./perf_initial_timeouts.png)
+
+We also see that many jobs are active for much longer than 1 second, even though the workers could complete them immediately.
```
Comment (Zelldon) [inline chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:131]: ❓ Per default the workers have a completion delay, did you change this?

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+When we scale up to 6 brokers, and thus distribute the partitions such that each broker is only leader for 1 instead of 2 partitions, we expect that processing performance improves.
+
+As these things usually go, we don't expect a doubling in performance but aiming for a 1.5x improvement seems reasonable.
+
+### Actual
+
+Shortly after scaling up and after partition leadership has balanced, we see a significant improvement in backpressure.
+![](./perf_after_backpressure.png)
+
+The job lifetime decreases dramatically, with most jobs now taking < 5ms from creation until completion.
```
Comment (Zelldon) [inline chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:152]: ❓ Isn;t this 50 ms? 🤔

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+This is a 1.6x improvement over the initial load while achieving similar backpressure.
+
+![](./perf_increased_load_backpressure.png)
+
+Job lifetime and overall processing latency is still better than before scaling up, even though load increased by 1.6x
+![](./perf_increased_load_job_lifetime.png)
+![](./perf_increased_load_processing_latency.png)
+
+
+Overall, this shows that scaling up can improve processing performance significantly, especially when the initial cluster setup is resource limited and vertical scaling is not possible.
```
Comment (Zelldon) [inline chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:183]: Awesome. Really nice results 😍

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/perf_after_cpu.png
Comment (Zelldon) [inline chaos-days/blog/2023-12-20-Broker-scaling-performance/perf_after_cpu.png:1]: ❓ Is there a specific reason why all images are duplicated as `-min`?

**Review by Zelldon — APPROVED** [review]
Awesome stuff @oleschoenburg and @deepthidevaki really great results congrats on this! Thank you very much for writing this down :)

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+2. Is scaling resilient to high processing load?
+3. Can scaling up improve processing performance?
+
+**TL;DR;** Scaling up works even under high load and has low impact on processing performance. After scaling is complete, processing performance improves in both throughput and latency.
+
+<!--truncate-->
+
+## Impact of scaling on processing performance
+
+Scaling up and down is an expensive operation where partition data is transferred between brokers, and leadership for partions changes.
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:29]: Normal raft replication

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md (hunk 1 above)
Comment (oleschoenburg) [reply chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:131]: Oh, I remember in the past the workers only used a completion delay if some "async" flag was set, which we did not do. Did that change, maybe with the job-push changes?

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/perf_after_cpu.png
Comment (oleschoenburg) [reply chaos-days/blog/2023-12-20-Broker-scaling-performance/perf_after_cpu.png:1]: Ohh, I wanted to compress them but thought it was in-place. Instead we now got all images twice :facepalm:

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md (hunk 1 above)
Comment (Zelldon) [reply chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:131]: No. This was always the case :)

File: chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md (hunk 2 above)
Comment (oleschoenburg) [reply chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md:152]: Ah damn, I _thought_ that 5ms is surprisingly good but didn't check it. Of course 50ms is correct :+1:

**Pull request merged**